### PR TITLE
WIP: rework metadata

### PR DESCRIPTION
**Pull request by miekg**
Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

**Comment by miekg** [comment]
This is (I think) where should head with this metadata thing.

this at least makes it type checked.

I've haven't fixed rewrite yet, because I won't to see if we can agree and where we're heading.

Regardless of implementation; it's pretty hard to understand what kind of variables are supported by the metadata plugin.

**Comment by johnbelamaric** [comment]
Yeah, this could be good. I want to be able to support ripping meta data fields out of EDNS0 options. This would become options on the metadata plugin, and would be at least functionally equivalent to what we can do here:

https://github.com/infobloxopen/themis/tree/master/contrib/coredns/policy#syntax

Those are typed as well. It reads in the data in one format and then provides the data in another format. So, incoming EDNS0 option could have data that is: raw bytes, a hex string, or an IP. We could extend that with an ordinary int too. Then the variable(s) that are extracted also get a type, currently only string and address are supported, and that plugin translates the inbound type to the outbound type. In our case we should decide if we want to convert to different types here or leave it to the consumer. If we leave it up to the consumer then "hex" probably isn't necessary - it's just a string and the consumer would have to know to treat it as hex.

File: plugin/metadata/provider.go [hunk 1]
```diff
@@ -11,16 +11,27 @@ import (
...
+	// List of variables which are provided by current Provider. Must not be changed during run-time.
+	VarInts() []string
+	VarStrings() []string
+
 	// Metadata is expected to return a value with metadata information by the key
 	// from 4th argument. Value can be later retrieved from context by any other plugin.
 	// If value is not available by some reason returned boolean value should be false.
-	Metadata(ctx context.Context, state request.Request, variable string) (interface{}, bool)
+	String(ctx context.Context, state request.Request, variable string) string
+	Int(ctx context.Context, state request.Request, variable string) int
```
Comment (fturib) [inline plugin/metadata/provider.go:22]: How will it work here when we find we would like a new type ? 
=> need to add 2 new functions to the interface Provider
Does it break the compatibility with all the already implemented plugin that will miss these 2 new interfaces ?
I guess yes ...

Would it be better, if we want to have typed metadata, to define enumeration of types and save the Name + Type + Value ?

And maybe the Metadata can offer the service to translate from Type to other Type of at least to storage (in the Getter/Setter)

Doing so, we keep only interface stable with only 2 functions:
- MetadataVars() => return array of (String, enumType)
- Metadata(ctx context.Context, state request.Request, variable string) (interface{}, bool) - here the type is already known.

and Metadata implement the convenient function for getting / setting the values. And these can be extended without changing the interface Provider.

What do you think ?

File: plugin/metadata/metadata.go [hunk 2]
```diff
@@ -31,9 +31,16 @@ func (m *Metadata) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
...
+			for _, v := range provider.VarStrings() {
+				val := provider.String(ctx, state, v)
+				if val != "" {
+					md.SetString(v, val)
+				}
+			}
+			for _, v := range provider.VarInts() {
+				val := provider.Int(ctx, state, v)
+				if val != -1 {
+					md.SetInt(v, val)
```
Comment (fturib) [inline plugin/metadata/metadata.go:43]: See other comment.

My fear is also here : for the plugins that need data, but do not care the type, it will have to loop over each type to get the right info ...

Comment (fturib) [reply plugin/metadata/metadata.go:43]: Although here we could implement a convenient function ... but that will come as the same as defining a enumType for the variable.

**Review by fturib — COMMENTED** [review]
I understand the point to have typed variables. But I was wondering on the lost of genericity,

and MOSTLY about compatibility when we want to implement a new type.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: rework m..." ]
>fturib commented on this pull request.
>
>I understand the point to have typed variables. But I was wondering on the lost of genericity,
>
>and MOSTLY about compatibility when we want to implement a new type.
>
>>  	// Metadata is expected to return a value with metadata information by the key
> 	// from 4th argument. Value can be later retrieved from context by any other plugin.
> 	// If value is not available by some reason returned boolean value should be false.
>-	Metadata(ctx context.Context, state request.Request, variable string) (interface{}, bool)
>+	String(ctx context.Context, state request.Request, variable string) string
>+	Int(ctx context.Context, state request.Request, variable string) int
>
>How will it work here when we find we would like a new type ?
>=> need to add 2 new functions to the interface Provider
>Does it break the compatibility with all the already implemented plugin that will miss these 2 new interfaces ?
>I guess yes ...
>
>Would it be better, if we want to have typed metadata, to define enumeration of types and save the Name + Type + Value ?
>
>And maybe the Metadata can offer the service to translate from Type to other Type of at least to storage (in the Getter/Setter)
>
>Doing so, we keep only interface stable with only 2 functions:
>- MetadataVars() => return array of (String, enumType)
>- Metadata(ctx context.Context, state request.Request, variable string) (interface{}, bool) - here the type is already known.
>
>and Metadata implement the convenient function for getting / setting the values. And these can be extended without changing the interface Provider.
>
>What do you think ?

Yeah, I think you make some good points. Currently the discoverability 
of the variables are supported is not that good. The variables package is also 
not sitting well with me, but I don't fully understand why I think that.
(Just filed https://github.com/coredns/coredns/issues/1911).

Basically, there are two things:

1. Metadata adds something to the metadata - this is trivial
2. Plugins need to retrieve all, or some data from the metadata (and need to 
   know the type of the data they get - this is not trivial

How does the plugin know what metadata is there and what type has it?

Say plugin X adds a `time.Time` to the metadata and names it 'epoch'.

Then plugin Y either knows fully about epoch and the time.Time type, or it can 
find out runtime. The variables package is an attempt to fix the latter, but 
then omitting the types. I think you need a run time registry, just like the 
plugin registration for new metadata variables or something.

**Comment by fturib** [comment]
About: 

> I think you need a run time registry, just like the
> plugin registration for new metadata variables or something.


In fact initial idea was only to collect ALL metadata that any plugin can provide about the incoming DNS query,  into a Map and let plugin Y (consumer) check what if what it needs is there ...

**idea is that we do not know what metadata we will get before the query exist.**
=> registration can happen ONLY query per query ....

introduction of initial "getMetadataName" is to allow a lazzy initialization for these metadata : at first your provide only the name .. and when one consumer NEED really that data, then we collect it. That is rather on optimization focus.

I think that "metadata" .. is all information we know about the query and its context.
the consumer cannot assume that metadata will always be there: it needs to check if this one exist or this one.
Example: we are thinking, for k8s, providing the labels attached to the pod or service. That would allow a plugin to filter the queries based on those labels.

There is kind of contract between Plugins that go by the configuration Corefile.
it is documentation on Plugin Y Consumer : I am able to process those metadata ... and they are usually provided by theses Plugin X, Z.

For the Type, we can keep simple - when we get the value, we get is into a standard format (let say a string or byte buffer), we also get the type and the plugin Metadata provides the convenient way to encode/decode the data from/to its original type.

We can get a small structure that embed the value encoded + type + way to decode it to standard types (string, int, ...)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: rework m..." ]
>About:
>
>> I think you need a run time registry, just like the
>> plugin registration for new metadata variables or something.
>
>In fact initial idea was only to collect ALL metadata that any plugin can provide about the incoming DNS query,  into a Map and let plugin Y (consumer) check what if what it needs is there ...

But duplicating data that's easily extracted (and much faster!) from the request 
and response writer is the wrong way to go.

>I think that "metadata" .. is all information we know about the query and its context.
>the consumer cannot assume that metadata will always be there: it needs to check if this one exist or this one.

Yes, but how? 

>Example: we are thinking, for k8s, providing the labels attached to the pod or service. That would allow a plugin to filter the queries based on those labels.

Yep, totally legit. This is, like, real metadata and impossible to get from the 
query (directly).

>There is kind of contract between Plugins that go by the configuration Corefile.
>it is documentation on Plugin Y Consumer : I am able to process those metadata ... and they are usually provided by theses Plugin X, Z.

That's a really brittle way of doing it. Is also weird: *adding* metdata is 
really neat; you just implement an interface; *Getting* the requires you reading 
some docs. Think such getters could be added to an structure that is:

map[string]func(...)

Where the key would be plugin's name adding the function. That function can then 
be called to return the value (if they are there).
[could be extended to a slice of functions)

>For the Type, we can keep simple - when we get the value, we get is into a standard format (let say a string or byte buffer), we also get the type and the plugin Metadata provides the convenient way to encode/decode the data from/to its original type.

Yep. I think string will be most useful.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: rework m..." ]
>About:
>
>> I think you need a run time registry, just like the
>> plugin registration for new metadata variables or something.

Just had another idea; what we are implementing here is basically
labelling the query. I.e. just like k8s labels. We could try to do it exactly 
like that.

I.e.

map[string]string

has the labels.

Labels() []string:  returns all keys (sorted?)
Value(label string) string: return the value for label

Yes, you still need to define the label name, but is does remove the need for 
variables.go. But also depends on reading the configuration on what labels are 
added by a plugin.

**Comment by miekg** [comment]
closing in favor of #1914
